### PR TITLE
Task 4235, rspconfig BMC dump for OpenBMC

### PR DESCRIPTION
**Pull request by xuweibj**
#4235 

* rspconfig -l|--list
```
[root@briggs01 dump]# rspconfig mid05tor12cn02 dump --list
mid05tor12cn02: [1] Elapsed: 09/22/2017 14:41:05, Size: 4624
mid05tor12cn02: [2] Elapsed: 09/22/2017 14:46:29, Size: 3608
mid05tor12cn02: [3] Elapsed: 10/05/2017 22:46:59, Size: 180976
mid05tor12cn02: [4] Elapsed: 10/09/2017 22:10:46, Size: 8184
mid05tor12cn02: [5] Elapsed: 10/09/2017 22:10:51, Size: 8404
mid05tor12cn02: [6] Elapsed: 10/09/2017 22:10:57, Size: 8384
mid05tor12cn02: [7] Elapsed: 10/26/2017 10:29:16, Size: 3692
mid05tor12cn02: [8] Elapsed: 10/26/2017 10:29:16, Size: 3716
mid05tor12cn02: [9] Elapsed: 10/26/2017 10:29:16, Size: 3760
mid05tor12cn02: [10] Elapsed: 10/26/2017 10:29:16, Size: 3676
```

* rspconfig -g|--generate
```
[root@briggs01 dump]# rspconfig mid05tor12cn02 dump -g
mid05tor12cn02: [11] success
```
When testing, found that need to wait several minutes could saw it in the list.

* rspconfig -c|--clear
```
[root@briggs01 dump]# rspconfig mid05tor12cn02 dump --clear 11
mid05tor12cn02: [11] clear
```

* rspconfig -d|--download
```
[root@briggs01 dump]# rspconfig mid05tor12cn02 dump --download 2
mid05tor12cn02: Saved dump 2 as /var/log/xcat/dump/mid05tor12cn02_dump_2.tar.xz
```
I modified some code in the same line with #4254, may has conflicts.

### CODE REVIEW

**Comment by whowutwut** [comment]
@xuweibj What happens if we issue `dump` by itself?  

Looking at usability scenario here.. I think most of the time we would want to call this to generate and download...  From the lessons learned in rflash... let me propose a use case here. 

`rspconfig <noderange> dump` 
   * Triggers a dump to be generated, the return of the create will give us the ID
   * Save the ID and then wait here for several minutes... 
   * When the ID is seen in the BMC, download it to the proposed directory 

The normal output could be.... 
```
Capturing BMC Dump information, this will take some time...
Generating Dump....
Downloading Dump...
```

If we have --verbose option, then we could do: 
```
Capturing BMC Dump information, this will take some time...
Generating BMC Dump....
Generating BMC Dump on <node>...
....
Downloading BMC Dump...
Downloading BMC Dump to /var/log/xcat/dump/<file>
```

The reason I state the above... is because the solution we implement does not scale or help the admin.  The above provides an xCAT interface into the REST but it does not help the user generate and capture Dumps across 100, 1000, 2000, 5000 nodes...  There's no way to guarantee the next dump ID is the same unless we do a clear before...  So we need to have the create pass back the ID, and then we have to wait until the ID is generated and capture the output...

**Comment by xuweibj** [comment]
@whowutwut , I want to make sure I understand your point well. Do you mean for command ``rspconfig dump``, we send a request to create a new one, and if could list it download it? So if do this, do we also need the options ``--list, --generate, --download``? Or is the requirement added the code for ``rspconfig dump`` itself in this pull request?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 1]
```diff
@@ -781,10 +816,22 @@ sub parse_args {
                 $setorget = "get";
             } elsif ($subcommand =~ /^sshcfg$/) {
                 $setorget = ""; # SSH Keys are copied using a RShellAPI, not REST API
+            } elsif ($subcommand eq "dump") {
+                my $option = $ARGV[1];
+                if ($option =~ /^-d$|^--download$/) {
+                    return ([ 1, "Invalid parameter for $command $option" ]) unless ($ARGV[2]);
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: The error msg is not making sense. 
What about 'No dump file ID specified'?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 2]
```diff
@@ -781,10 +816,22 @@ sub parse_args {
...
+                    return ([ 1, "Invalid parameter for $command $option $ARGV[2]" ]) if ($ARGV[2] !~ /^\d*$/);
+                } elsif ($option =~ /^-c$|^--clear$/) {
+                    return ([ 1, "Invalid parameter for $command $option" ]) unless ($ARGV[2]);
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Same comment with line 822.

**Comment by zet809** [comment]
hi, @whowutwut , you are right that this interface may not make sense for groups of nodes, we need to provide a better interface quickly.
But maybe sometime, if user want to download a dump file days before, he can use the interface right now.(Just as `rflash <node> <image_id> -a` to active a specified version of firmware, right?)
So, I think we can merge it right now.

**Pull request merged**